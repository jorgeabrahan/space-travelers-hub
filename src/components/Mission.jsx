import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import styles from '../css/Missions.module.css';
import { bookMission, leaveMission } from '../redux/missions/missionsSlice';

function Mission(
  {
    missionId,
    missionName,
    description,
    odd,
    reserved,
  },
) {
  const dispatch = useDispatch();
  return (
    <tr className={odd ? styles.tableOdd : styles.tableEven}>
      <td className={styles.tableMissionName}>{missionName}</td>
      <td className={styles.tableMissionDescription}>{description}</td>
      <td className={styles.tableStatus}>
        {
          odd ? (
            <div className={styles.notAMember}>not a member</div>
          ) : (
            <div className={styles.member}>active member</div>
          )
        }
      </td>
      <td className={styles.tableJoin}>
        {
          reserved ? (
            <button
              className={styles.onMission}
              type="button"
              onClick={() => { dispatch(leaveMission(missionId)); }}
            >
              Leave Mission
            </button>
          ) : (
            <button
              className={styles.outMission}
              type="button"
              onClick={() => { dispatch(bookMission(missionId)); }}
            >
              Join Mission
            </button>
          )
        }
      </td>
    </tr>
  );
}

Mission.propTypes = {
  missionName: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  odd: PropTypes.bool.isRequired,
  reserved: PropTypes.bool.isRequired,
  missionId: PropTypes.string.isRequired,
};

export default Mission;
