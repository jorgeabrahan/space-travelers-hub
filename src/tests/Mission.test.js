import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import Mission from '../components/Mission';
import store from '../redux/store';

describe('Tests for <Mission /> component', () => {
  const title = 'Mission 1';
  const description = 'Mission description';
  const createMission = (reserved = false) => (
    <Mission
      missionId={'2'}
      missionName={title}
      description={description}
      odd={true}
      reserved={reserved}
    />
  );
  const wrapper = (component) => (
    <Provider store={store}>
      <table>
        <tbody>
          {component}
        </tbody>
      </table>
    </Provider>
  );
  it('Should render the content correctly', () => {
    render(wrapper(createMission()));
    expect(screen.getByText(title)).not.toBeNull();
    expect(screen.getByText('Mission description')).not.toBeNull();
  });

  it('Should display the reserve button when reserved is false', () => {
    render(wrapper(createMission()));
    expect(screen.queryByRole('button', { children: 'not a member' })).not.toBeNull();
    expect(() => screen.getByText('Leave Mission')).toThrow();
  });

  it('Should display the cancel reservation button when reserved is true', () => {
    render(wrapper(createMission(true)));
    expect(screen.queryByRole('button', { children: 'active member' })).not.toBeNull();
    expect(() => screen.getByText('Join Mission')).toThrow();
  });

  it('Should match the snapshot', () => {
    expect(renderer.create(wrapper(createMission())).toJSON()).toMatchSnapshot();
  });
});