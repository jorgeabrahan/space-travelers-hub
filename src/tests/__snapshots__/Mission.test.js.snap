// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <Mission /> component Should match the snapshot 1`] = `
<table>
  <tbody>
    <tr
      className="tableOdd"
    >
      <td
        className="tableMissionName"
      >
        Mission 1
      </td>
      <td
        className="tableMissionDescription"
      >
        Mission description
      </td>
      <td
        className="tableStatus"
      >
        <div
          className="notAMember"
        >
          not a member
        </div>
      </td>
      <td
        className="tableJoin"
      >
        <button
          className="outMission"
          onClick={[Function]}
          type="button"
        >
          Join Mission
        </button>
      </td>
    </tr>
  </tbody>
</table>
`;
